File: backend/app/api/v1/watchlist.py
"""Watchlist endpoints (自选股) — all require auth, isolated by user_id."""

from __future__ import annotations

from fastapi import APIRouter, Depends

from app.api.deps import get_current_user
from app.core.response import success
from app.models.user import User
from app.schemas.watchlist import AddWatchlistRequest, UpdateWatchlistRequest
from app.services import watchlist

router = APIRouter()


@router.get("")
def get_watchlist(user: User = Depends(get_current_user)) -> dict:
    return success(watchlist.list_items(str(user.id)))


@router.get("/groups")
def get_groups(user: User = Depends(get_current_user)) -> dict:
    return success(watchlist.list_groups(str(user.id)))


@router.post("")
def add(body: AddWatchlistRequest, user: User = Depends(get_current_user)) -> dict:
    return success(watchlist.add_item(str(user.id), body.code, body.name, body.group))


@router.patch("/{code}")
def patch(
    code: str, body: UpdateWatchlistRequest, user: User = Depends(get_current_user)
) -> dict:
    ok = watchlist.update_item(str(user.id), code, body.group, body.sortOrder)
    return success({"updated": ok})


@router.delete("/{code}")
def remove(code: str, user: User = Depends(get_current_user)) -> dict:
    ok = watchlist.remove_item(str(user.id), code)
    return success({"removed": ok})

File: frontend/src/lib/api/watchlist.ts
import { apiClient } from './client';

export interface WatchlistItemView {
  code: string;
  name: string;
  group: string;
  sortOrder: number;
  price: number | null;
  change: number | null;
  changePercent: number | null;
  createdAt: string | null;
}

/**
 * 自选股 API（均需登录，按 user_id 隔离）。
 */
export const watchlistApi = {
  getList: async (): Promise<WatchlistItemView[]> => {
    const res = await apiClient.get<WatchlistItemView[]>('/watchlist');
    return res.data;
  },

  getGroups: async (): Promise<string[]> => {
    const res = await apiClient.get<string[]>('/watchlist/groups');
    return res.data;
  },

  add: async (
    code: string,
    options?: { name?: string; group?: string }
  ): Promise<{ code: string; added: boolean }> => {
    const res = await apiClient.post<{ code: string; added: boolean }>('/watchlist', {
      code,
      name: options?.name ?? '',
      group: options?.group ?? '默认分组',
    });
    return res.data;
  },

  update: async (
    code: string,
    body: { group?: string; sortOrder?: number }
  ): Promise<{ updated: boolean }> => {
    const res = await apiClient.patch<{ updated: boolean }>(
      `/watchlist/${encodeURIComponent(code)}`,
      body
    );
    return res.data;
  },

  remove: async (code: string): Promise<{ removed: boolean }> => {
    const res = await apiClient.delete<{ removed: boolean }>(
      `/watchlist/${encodeURIComponent(code)}`
    );
    return res.data;
  },
};
